modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTable.java

```diff
@@ -0,0 +1,305 @@
+import io.siddhi.query.api.expression.condition.And;
+import io.siddhi.query.api.expression.condition.Compare;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static io.siddhi.query.api.util.AnnotationHelper.getAnnotation;
+
+/**
+ * common interface for FIFO, LRU, and LFU cache tables
+ */
+public abstract class CacheTable extends InMemoryTable {
+    private int maxSize;
+    private boolean cacheExpiryEnabled;
+    SiddhiAppContext siddhiAppContext;
+    int cachePolicyAttributePosition;
+    int numColumns;
+    int expiryAttributePosition;
+
+    @Override
+    public void init(TableDefinition tableDefinition, StreamEventFactory storeEventPool,
+                     StreamEventCloner storeEventCloner, ConfigReader configReader, SiddhiAppContext siddhiAppContext,
+                     RecordTableHandler recordTableHandler) {
+        this.tableDefinition = tableDefinition;
+        this.tableStreamEventCloner = storeEventCloner;
+        EventHolder eventHolder = EventHolderPasser.parse(tableDefinition, storeEventPool, siddhiAppContext, true);
+
+        if (eventHolder instanceof IndexEventHolderForCache) {
+            ((IndexEventHolderForCache) eventHolder).setCacheTable(this);
+        }
+
+        stateHolder = siddhiAppContext.generateStateHolder(tableDefinition.getId(),
+                () -> new TableState(eventHolder));
+    }
+
+    public void initCacheTable(TableDefinition cacheTableDefinition, ConfigReader configReader,
+                               SiddhiAppContext siddhiAppContext, RecordTableHandler recordTableHandler,
+                               boolean cacheExpiryEnabled, int maxSize, String cachePolicy) {
+        this.maxSize = maxSize;
+        this.cacheExpiryEnabled = cacheExpiryEnabled;
+        this.siddhiAppContext = siddhiAppContext;
+        addRequiredFieldsToCacheTableDefinition(cacheTableDefinition, cacheExpiryEnabled);
+
+        // initialize cache table
+        MetaStreamEvent cacheTableMetaStreamEvent = new MetaStreamEvent();
+        cacheTableMetaStreamEvent.addInputDefinition(cacheTableDefinition);
+        for (Attribute attribute : cacheTableDefinition.getAttributeList()) {
+            cacheTableMetaStreamEvent.addOutputData(attribute);
+        }
+
+        StreamEventFactory cacheTableStreamEventFactory = new StreamEventFactory(cacheTableMetaStreamEvent);
+        StreamEventCloner cacheTableStreamEventCloner = new StreamEventCloner(cacheTableMetaStreamEvent,
+                cacheTableStreamEventFactory);
+        super.initTable(cacheTableDefinition, cacheTableStreamEventFactory,
+                cacheTableStreamEventCloner, configReader, siddhiAppContext, recordTableHandler);
+    }
+
+    public void addStreamEventUptoMaxSize(StreamEvent streamEvent) {
+        int sizeAfterAdding = this.size();
+        ComplexEventChunk<StreamEvent> addEventsLimitCopy = new ComplexEventChunk<>(true);
+        do {
```

Comment:
Use while loop interest of do-while loop as this can eliminate the assumption of first event is not null